url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/PricingPolicy.java

```diff
@@ -0,0 +1,457 @@
+		 * @param aenergyPricing the price of energy consumed
+		 * @param apowerPricing the power pricing of the contracted capacity
+		 * @return the builder
+		 */
+		public Builder energyPowerPricing(double acontractedCapacity, double aenergyPricing, double apowerPricing) {
+			if (!type.equals("EnergyPowerPricing"))
+			{
+				System.err.println("PricingPolicy type mismatch: method only applicable to the EnergyPowerPricing scheme");
+				return this;
+			}
+			contractedCapacity = acontractedCapacity;
+			energyPricing = aenergyPricing;
+			powerPricing = apowerPricing;
+			return this;
+		}	
+
+		/**
+		 * Sets additional parameters for an "All Inclusive Pricing" policy.
+		 *
+		 * @param afixedCost the afixed cost
+		 * @param aadditionalCost the aadditional cost
+		 * @param acontractedEnergy the acontracted energy
+		 * @return the builder
+		 */
+		public Builder allInclusivePricing(double afixedCost, double aadditionalCost, double acontractedEnergy) {
+			if (!type.equals("AllInclusivePricing"))
+			{
+				System.err.println("PricingPolicy type mismatch: method only applicable to the AllInclusivePricing scheme");
+				return this;
+			}
+			contractedEnergy = acontractedEnergy;
+			fixedCost = afixedCost;
+			additionalCost = aadditionalCost;
+			return this;
+		}
+
+		/**
+		 * Sets additional parameters for an "TOU (Time Of Use) Pricing" policy.
+		 *
+		 * @param froms the froms
+		 * @param tos the tos
+		 * @param prices the prices
+		 * @return the builder
+		 */
+		public Builder touPricing(String[] froms, String[] tos, double[] prices) {
+			if (!type.equals("TOUPricing"))
+			{
+				System.err.println("PricingPolicy type mismatch: method only applicable to the TOUPricing scheme");
+				return this;
+			}
+			if (! (froms.length == tos.length && tos.length == prices.length))
+			{
+				System.err.println("PricingPolicy initialization error: all input tables must have the same length");
+				return this;
+			}
+
+			for(int i = 0; i < froms.length; i++) {
+				String from = froms[i];
+				String to = tos[i];
+				double price = prices[i];
+				Period p = new Period(from, to, price);
+				periods.add(p);
+			}
+			return this;
+		}
+
+		/**
+		 * Sets additional parameters for an "Scalar Energy Pricing" policy.
+		 *
+		 * @param prices the prices
+		 * @param alevels the alevels
+		 * @return the builder
+		 */
+		public Builder scalarEnergyPricing(double[] prices, double[] alevels) {
+			if (!type.equals("ScalarEnergyPricing"))
+			{
+				System.err.println("PricingPolicy type mismatch: method only applicable to the ScalarEnergyPricing scheme");
+				return this;
+			}
+			if (! (alevels.length == prices.length))
+			{
+				System.err.println("PricingPolicy initialization error: all input tables must have the same length");
+				return this;
+			}
+
+			for(int i = 0; i < prices.length; i++) {
+				double price = prices[i];
+				double level = alevels[i];
+				Level l = new Level(price, level);
+				levels.add(l);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 208:213, 240:245.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.